llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -761,5 +761,147 @@ getLargeGRFSplitter(ModuleDesc &&MD, bool EmitOnlyKernelsAsEntryPoints) {
+        Aspects.push_back(ExtractIntegerFromMDNodeOperand(MDN, I));
+      }
+      llvm::sort(Aspects);
+    }
+
+    llvm::hash_code AspectsHash =
+        llvm::hash_combine_range(Aspects.begin(), Aspects.end());
+    Hash = static_cast<unsigned>(llvm::hash_combine(AspectsHash));
+  }
+
+  std::string getName(StringRef BaseName) const {
+    if (Aspects.empty())
+      return BaseName.str() + "-no-aspects";
+
+    std::string Ret = BaseName.str() + "-aspects";
+    for (int A : Aspects) {
+      Ret += "-" + std::to_string(A);
+    }
+    return Ret;
+  }
+
+  static UsedOptionalFeatures getTombstone() {
+    UsedOptionalFeatures Ret;
+    Ret.IsTombstoneKey = true;
+    return Ret;
+  }
+
+  static UsedOptionalFeatures getEmpty() {
+    UsedOptionalFeatures Ret;
+    Ret.IsEmpty = true;
+    return Ret;
+  }
+
+private:
+  // For DenseMap:
+  llvm::hash_code Hash = {};
+  bool IsTombstoneKey = false;
+  bool IsEmpty = false;
+
+public:
+  bool operator==(const UsedOptionalFeatures &Other) const {
+    // Tombstone does not compare equal to any other item
+    if (IsTombstoneKey || Other.IsTombstoneKey)
+      return false;
+
+    if (Aspects.size() != Other.Aspects.size())
+      return false;
+
+    for (size_t I = 0, E = Aspects.size(); I != E; ++I) {
+      if (Aspects[I] != Other.Aspects[I])
+        return false;
+    }
+
+    return IsEmpty == Other.IsEmpty;
+  }
+
+  unsigned hash() const { return static_cast<unsigned>(Hash); }
+};
+
+struct UsedOptionalFeaturesAsKeyInfo {
+  static inline UsedOptionalFeatures getEmptyKey() {
+    return UsedOptionalFeatures::getEmpty();
+  }
+
+  static inline UsedOptionalFeatures getTombstoneKey() {
+    return UsedOptionalFeatures::getTombstone();
+  }
+
+  static unsigned getHashValue(const UsedOptionalFeatures &Value) {
+    return Value.hash();
+  }
+
+  static bool isEqual(const UsedOptionalFeatures &LHS,
+                      const UsedOptionalFeatures &RHS) {
+    return LHS == RHS;
+  }
+};
+} // namespace
+
+std::unique_ptr<ModuleSplitterBase>
+getSplitterByOptionalFeatures(ModuleDesc &&MD,
+                              bool EmitOnlyKernelsAsEntryPoints) {
+  EntryPointGroupVec Groups;
+
+  DenseMap<UsedOptionalFeatures, EntryPointSet, UsedOptionalFeaturesAsKeyInfo>
+      PropertiesToFunctionsMap;
+
+  Module &M = MD.getModule();
+
+  // Only process module entry points:
+  for (auto &F : M.functions()) {
+    if (!isEntryPoint(F, EmitOnlyKernelsAsEntryPoints) ||
+        !MD.isEntryPointCandidate(F)) {
+      continue;
+    }
+
+    auto Key = UsedOptionalFeatures(&F);
+    PropertiesToFunctionsMap[std::move(Key)].insert(&F);
+  }
+
+  if (!PropertiesToFunctionsMap.empty()) {
+    Groups.reserve(PropertiesToFunctionsMap.size());
+    for (auto &EPG : PropertiesToFunctionsMap) {
+      Groups.emplace_back(EntryPointGroup{
+          EPG.first.getName(MD.getEntryPointGroup().GroupId),
+          std::move(EPG.second), MD.getEntryPointGroup().Props});
```

Comment:
I can't be sure with all those type aliases used, but are we forwarding the args to the actual ctor, or is copy/move ctor involved here?

Reply:
Indeed, we are creating an object first and then passing it as a whole to `emplace_back` which uses default copy/move constructor. I will send a separate PR to adjust other uses of `emplace_back`.

Fixed this one in 0cc45b2c1a04551249551bf104bd99b559a55d83

Reply:
See #7469 for the rest of `emplace_back` occurrences